### PR TITLE
feat: port hard constraints for type check from Draco1

### PR DESCRIPTION
**Pull request by Zehua-Zeng**

### CODE REVIEW

File: draco/asp/hard.lp [hunk 1]
```diff
@@ -141,3 +141,54 @@ violation(view_scale_conflict) :-
     entity(scale,V,S2),
     attribute((scale,channel),S1,C),
     attribute((scale,channel),S2,C).
+
+% @hard(invalid_bin) Check bin type.
+violation(invalid_bin) :-
+    attribute((encoding,binning),_,B),
+    not B >= 0.
```
Comment (domoritz) [inline draco/asp/hard.lp:148]:
```suggestion
    B < 0.
```

File: draco/asp/hard.lp [hunk 2]
```diff
@@ -141,3 +141,54 @@ violation(view_scale_conflict) :-
...
+
+% @hard(invalid_num_rows) Num_rows has to be larger than 0.
```
Comment (domoritz) [inline draco/asp/hard.lp:150]:
```suggestion
% @hard(invalid_num_rows) number_rows has to be larger than 0.
```

File: draco/asp/hard.lp [hunk 3]
```diff
@@ -141,3 +141,54 @@ violation(view_scale_conflict) :-
...
+violation(invalid_unique) :-
+    attribute((field,unique),_,U),
+    U <= 0.
+
+% @hard(invalid_extent_non_number_min) Extent only allowed for numbers (for now).
+violation(invalid_extent_non_number_min) :-
+    attribute((field,min),F,_),
+    not attribute((field,type),F,number).
+
+% @hard(invalid_extent_non_number_max) Extent only allowed for numbers (for now).
```
Comment (domoritz) [inline draco/asp/hard.lp:165]: Let's merge this with the rule definition above.

File: draco/asp/hard.lp [hunk 4]
```diff
@@ -141,3 +141,54 @@ violation(view_scale_conflict) :-
...
+    attribute((field,min),F,MIN),
+    attribute((field,max),F,MAX),
+    MIN > MAX.
+
+% @hard(invalid_non_string_freq) Frequency for strings only.
+violation(invalid_non_string_freq) :-
+    attribute((field,freq),F,_),
+    not attribute((field,type),F,string).
+
+% @hard(encoding_field_same_name) The name of a field cannot be the name of an encoding. This is to prevent errors coming from the shortcuts in define.lp.
```
Comment (domoritz) [inline draco/asp/hard.lp:191]: I don't understand this comment. Can you point to the problematic rule?

**Review by domoritz — CHANGES_REQUESTED** [review]

File: draco/asp/hard.lp (hunk 4 above)
Comment (Zehua-Zeng) [reply draco/asp/hard.lp:191]: https://github.com/uwdata/draco/blob/master/asp/hard.lp#:~:text=%25%20%40constraint%20The%20name,N)%2C%20field(N).

**Comment by Zehua-Zeng** [comment]
I made changes based on all requests

**Comment by domoritz** [comment]
Looks great. Please update the pull request so we can merge it. (merge main into this).

File: draco/tests/test_hard.py [hunk 5]
```diff
@@ -1381,3 +1381,262 @@ def test_aggregate_not_all_continuous():
...
+    """
+    )
+
+    assert no_violations(
+        p
+        + """
+    entity(encoding,1,2).
+    attribute((encoding,channel),2,x).
+    attribute((encoding,field),2,temperature).
+    attribute((encoding,binning),2).
```
Comment (JunranY) [inline draco/tests/test_hard.py:1407]: I think we now require the num of bins value based on previous discussion.

**Review by JunranY — APPROVED** [review]

File: draco/tests/test_hard.py (hunk 5 above)
Comment (JunranY) [reply draco/tests/test_hard.py:1407]: @domoritz

Comment (domoritz) [reply draco/tests/test_hard.py:1407]: I think that makes sense. Is this okay here?

File: draco/tests/test_hard.py [hunk 6]
```diff
@@ -1381,3 +1381,262 @@ def test_aggregate_not_all_continuous():
...
+    """
+    )
+
+    assert (
```
Comment (domoritz) [inline draco/tests/test_hard.py:1411]: Add a comment explaining the issue in each of the examples below.

**Review by domoritz — APPROVED** [review]

File: draco/tests/test_hard.py (hunk 5 above)
Comment (Zehua-Zeng) [reply draco/tests/test_hard.py]: deleted the test

File: draco/tests/test_hard.py (hunk 6 above)
Comment (Zehua-Zeng) [reply draco/tests/test_hard.py:1402]: added comments

**Pull request merged**